src/graph/Graph.java

```diff
@@ -87,6 +88,29 @@ public void clear() {
   }
 
   @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || this.getClass() != o.getClass()) {
+      return false;
+    }
+
+    Graph g = (Graph) o;
+    if (g.getNumOfVertices() != this.getNumOfVertices()) {
+      return false;
+    }
+
+    for (Vertex u : this.getVertices()) {
```

Comment:
Please, put this.getVertices out of for loop.

Reply:
I also found some other places that has this.getVertices etc in the loop. Please fix.